File: packages/ts-moose-lib/src/dmv2/sdk/workflow.ts
import { getMooseInternal } from "../internal";

/**
 * Context passed to task handlers. Single param to future-proof API changes.
 *
 * - state: shared mutable state for the task and its lifecycle hooks
 * - input: optional typed input for the task (undefined when task has no input)
 */
/**
 * Task handler context. If the task declares an input type (T != null),
 * `input` is required and strongly typed. For no-input tasks (T = null),
 * `input` is omitted/optional.
 */
export type TaskContext<TInput> =
  TInput extends null ? { state: any; input?: null }
  : { state: any; input: TInput };

/**
 * Configuration options for defining a task within a workflow.
 *
 * @template T - The input type for the task
 * @template R - The return type for the task
 */
export interface TaskConfig<T, R> {
  /** The main function that executes the task logic */
  run: (context: TaskContext<T>) => Promise<R>;

  /**
   * Optional array of tasks to execute after this task completes successfully.
   * Supports all combinations of input types (real type or null) and output types (real type or void).
   * When this task returns void, onComplete tasks expect null as input.
   * When this task returns a real type, onComplete tasks expect that type as input.
   */
  onComplete?: (
    | Task<R extends void ? null : R, any>
    | Task<R extends void ? null : R, void>
  )[];

  /**
   * Optional function that is called when the task is cancelled.
   */
  /** Optional function that is called when the task is cancelled. */
  onCancel?: (context: TaskContext<T>) => Promise<void>;

  /** Optional timeout duration for the task execution (e.g., "30s", "5m") */
  timeout?: string;

  /** Optional number of retry attempts if the task fails */
  retries?: number;
}

/**
 * Represents a single task within a workflow system.
 *
 * A Task encapsulates the execution logic, completion handlers, and configuration
 * for a unit of work that can be chained with other tasks in a workflow.
 *
 * @template T - The input type that this task expects
 * @template R - The return type that this task produces
 */
export class Task<T, R> {
  /**
   * Creates a new Task instance.
   *
   * @param name - Unique identifier for the task
   * @param config - Configuration object defining the task behavior
   *
   * @example
   * ```typescript
   * // No input, no output
   * const task1 = new Task<null, void>("task1", {
   *   run: async () => {
   *     console.log("No input/output");
   *   }
   * });
   *
   * // No input, but has output
   * const task2 = new Task<null, OutputType>("task2", {
   *   run: async () => {
   *     return someOutput;
   *   }
   * });
   *
   * // Has input, no output
   * const task3 = new Task<InputType, void>("task3", {
   *   run: async (input: InputType) => {
   *     // process input but return nothing
   *   }
   * });
   *
   * // Has both input and output
   * const task4 = new Task<InputType, OutputType>("task4", {
   *   run: async (input: InputType) => {
   *     return process(input);
   *   }
   * });
   * ```
   */
  constructor(
    readonly name: string,
    readonly config: TaskConfig<T, R>,
  ) {}
}

/**
 * Configuration options for defining a workflow.
 *
 * A workflow orchestrates the execution of multiple tasks in a defined sequence
 * or pattern, with support for scheduling, retries, and timeouts.
 */
export interface WorkflowConfig {
  /**
   * The initial task that begins the workflow execution.
   * Supports all combinations of input types (real type or null) and output types (real type or void):
   * - Task<null, OutputType>: No input, returns a type
   * - Task<null, void>: No input, returns nothing
   * - Task<InputType, OutputType>: Has input, returns a type
   * - Task<InputType, void>: Has input, returns nothing
   */
  startingTask:
    | Task<null, any>
    | Task<null, void>
    | Task<any, any>
    | Task<any, void>;

  /** Optional number of retry attempts if the entire workflow fails */
  retries?: number;

  /** Optional timeout duration for the entire workflow execution (e.g., "10m", "1h") */
  timeout?: string;

  /** Optional cron-style schedule string for automated workflow execution */
  schedule?: string;
}

/**
 * Represents a complete workflow composed of interconnected tasks.
 *
 * A Workflow manages the execution flow of multiple tasks, handling scheduling,
 * error recovery, and task orchestration. Once created, workflows are automatically
 * registered with the internal Moose system.
 *
 * @example
 * ```typescript
 * const dataProcessingWorkflow = new Workflow("dataProcessing", {
 *   startingTask: extractDataTask,
 *   schedule: "0 2 * * *", // Run daily at 2 AM
 *   timeout: "1h",
 *   retries: 2
 * });
 * ```
 */
export class Workflow {
  /**
   * Creates a new Workflow instance and registers it with the Moose system.
   *
   * @param name - Unique identifier for the workflow
   * @param config - Configuration object defining the workflow behavior and task orchestration
   * @throws {Error} When the workflow contains null/undefined tasks or infinite loops
   */
  constructor(
    readonly name: string,
    readonly config: WorkflowConfig,
  ) {
    const workflows = getMooseInternal().workflows;
    if (workflows.has(name)) {
      throw new Error(`Workflow with name ${name} already exists`);
    }
    this.validateTaskGraph(config.startingTask, name);
    workflows.set(name, this);
  }

  /**
   * Validates the task graph to ensure there are no null tasks or infinite loops.
   *
   * @private
   * @param startingTask - The starting task to begin validation from
   * @param workflowName - The name of the workflow being validated (for error messages)
   * @throws {Error} When null/undefined tasks are found or infinite loops are detected
   */
  private validateTaskGraph(
    startingTask: Task<any, any> | null | undefined,
    workflowName: string,
  ): void {
    if (startingTask === null || startingTask === undefined) {
      throw new Error(
        `Workflow "${workflowName}" has a null or undefined starting task`,
      );
    }

    const visited = new Set<string>();
    const recursionStack = new Set<string>();

    const validateTask = (
      task: Task<any, any> | null | undefined,
      currentPath: string[],
    ): void => {
      if (task === null || task === undefined) {
        const pathStr =
          currentPath.length > 0 ? currentPath.join(" -> ") + " -> " : "";
        throw new Error(
          `Workflow "${workflowName}" contains a null or undefined task in the task chain: ${pathStr}null`,
        );
      }

      const taskName = task.name;

      if (recursionStack.has(taskName)) {
        const cycleStartIndex = currentPath.indexOf(taskName);
        const cyclePath =
          cycleStartIndex >= 0 ?
            currentPath.slice(cycleStartIndex).concat(taskName)
          : currentPath.concat(taskName);
        throw new Error(
          `Workflow "${workflowName}" contains an infinite loop in task chain: ${cyclePath.join(" -> ")}`,
        );
      }

      if (visited.has(taskName)) {
        // Already processed this task and its children
        return;
      }

      visited.add(taskName);
      recursionStack.add(taskName);

      if (task.config.onComplete) {
        for (const nextTask of task.config.onComplete) {
          validateTask(nextTask, [...currentPath, taskName]);
        }
      }

      recursionStack.delete(taskName);
    };

    validateTask(startingTask, []);
  }
}

File: packages/py-moose-lib/moose_lib/dmv2/workflow.py
"""
Workflow definitions for Moose Data Model v2 (dmv2).

This module provides classes for defining and configuring workflows composed of tasks,
including task dependencies, configurations, and execution functions.
"""
import dataclasses
from typing import Any, Optional, Dict, List, Callable, Union, Awaitable, Generic
from pydantic import BaseModel

from .types import TypedMooseResource, T_none, U_none
from ._registry import _workflows

type TaskRunFunc[T_none, U_none] = Union[
    # Case 1: No input, no output
    Callable[[], None],
    # Case 2: No input, with output
    Callable[[], Union[U_none, Awaitable[U_none]]],
    # Case 3: With input, no output
    Callable[[T_none], None],
    # Case 4: With input, with output
    Callable[[T_none], Union[U_none, Awaitable[U_none]]]
]

@dataclasses.dataclass
class TaskConfig(Generic[T_none, U_none]):
    """Configuration for a Task.

    Attributes:
        run: The handler function that executes the task logic.
        on_complete: Optional list of tasks to run after this task completes.
        timeout: Optional timeout string (e.g. "5m", "1h").
        retries: Optional number of retry attempts.
    """
    run: TaskRunFunc[T_none, U_none]
    on_complete: Optional[list["Task[U_none, Any]"]] = None
    timeout: Optional[str] = None
    retries: Optional[int] = None

class Task(TypedMooseResource, Generic[T_none, U_none]):
    """Represents a task that can be executed as part of a workflow.

    Tasks are the basic unit of work in a workflow, with typed input and output.
    They can be chained together using the on_complete configuration.

    Args:
        name: The name of the task.
        config: Configuration specifying the task's behavior.
        t: The Pydantic model defining the task's input schema
           (passed via `Task[InputModel, OutputModel](...)`).
           OutputModel can be None for tasks that don't return a value.

    Attributes:
        config (TaskConfig[T, U]): The configuration for this task.
        columns (Columns[T]): Helper for accessing input field names safely.
        name (str): The name of the task.
        model_type (type[T]): The Pydantic model associated with this task's input.
    """
    config: TaskConfig[T_none, U_none]

    def __init__(self, name: str, config: TaskConfig[T_none, U_none], **kwargs):
        super().__init__()
        self._set_type(name, self._get_type(kwargs))
        self.config = config

    @classmethod
    def _get_type(cls, keyword_args: dict):
        t = keyword_args.get('t')
        if t is None:
            raise ValueError(f"Use `{cls.__name__}[T, U](name='...')` to supply both input and output types")
        if not isinstance(t, tuple) or len(t) != 2:
            raise ValueError(f"Use `{cls.__name__}[T, U](name='...')` to supply both input and output types")

        input_type, output_type = t
        if input_type is not None and (not isinstance(input_type, type) or not issubclass(input_type, BaseModel)):
            raise ValueError(f"Input type {input_type} is not a Pydantic model or None")
        if output_type is not None and (not isinstance(output_type, type) or not issubclass(output_type, BaseModel)):
            raise ValueError(f"Output type {output_type} is not a Pydantic model or None")
        return t

    def _set_type(self, name: str, t: tuple[type[T_none], type[U_none]]):
        input_type, output_type = t
        self._t = input_type
        self._u = output_type
        self.name = name

@dataclasses.dataclass
class WorkflowConfig:
    """Configuration for a workflow.

    Attributes:
        starting_task: The first task to execute in the workflow.
        retries: Optional number of retry attempts for the entire workflow.
        timeout: Optional timeout string for the entire workflow.
        schedule: Optional cron-like schedule string for recurring execution.
    """
    starting_task: Task[Any, Any]
    retries: Optional[int] = None
    timeout: Optional[str] = None
    schedule: Optional[str] = None

class Workflow:
    """Represents a workflow composed of one or more tasks.

    Workflows define a sequence of tasks to be executed, with optional
    scheduling, retries, and timeouts at the workflow level.

    Args:
        name: The name of the workflow.
        config: Configuration specifying the workflow's behavior.

    Attributes:
        name (str): The name of the workflow.
        config (WorkflowConfig): The configuration for this workflow.
    """
    def __init__(self, name: str, config: WorkflowConfig):
        self.name = name
        self.config = config
        # Register the workflow in the internal registry
        _workflows[name] = self

    def get_task_names(self) -> list[str]:
        """Get a list of all task names in this workflow.

        Returns:
            list[str]: List of task names in the workflow, including all child tasks
        """
        def collect_task_names(task: Task) -> list[str]:
            names = [task.name]
            if task.config.on_complete:
                for child in task.config.on_complete:
                    names.extend(collect_task_names(child))
            return names

        return collect_task_names(self.config.starting_task)

    def get_task(self, task_name: str) -> Optional[Task]:
        """Find a task in this workflow by name.

        Args:
            task_name: The name of the task to find

        Returns:
            Optional[Task]: The task if found, None otherwise
        """
        def find_task(task: Task) -> Optional[Task]:
            if task.name == task_name:
                return task
            if task.config.on_complete:
                for child in task.config.on_complete:
                    found = find_task(child)
                    if found:
                        return found
            return None

        return find_task(self.config.starting_task)